src/Utils.vala

```diff
@@ -117,4 +118,140 @@ namespace Utils {
+        try {
+            FileUtils.get_contents ("/etc/machine-id", out machine_id);
+        } catch (FileError e) {
+            warning ("%s", e.message);
+            return null;
+        }
+
+        return machine_id.strip ();
+    }
+
+    private static string? get_sys_vendor () {
+        string vendor;
+        try {
+            FileUtils.get_contents ("/sys/devices/virtual/dmi/id/sys_vendor", out vendor);
+        } catch (FileError e) {
+            warning ("%s", e.message);
+            return null;
+        }
+
+        return vendor.strip ();
+    }
+
+    private static string? get_product_name () {
+        string model;
+        try {
+            FileUtils.get_contents ("/sys/devices/virtual/dmi/id/product_name", out model);
+        } catch (FileError e) {
+            warning ("%s", e.message);
+            return null;
+        }
+
+        return model.strip ();
+    }
+
+    private static string? get_product_version () {
+        string model;
+        try {
+            FileUtils.get_contents ("/sys/devices/virtual/dmi/id/product_version", out model);
+        } catch (FileError e) {
+            warning ("%s", e.message);
+            return null;
+        }
+
+        return model.strip ();
+    }
+
+    // Based on https://git.launchpad.net/ubiquity/tree/ubiquity/misc.py?id=ae6415d224c2e76afa2274cc9f85997f38870419#n648
+    private static string? get_ubiquity_compatible_hostname () {
+        string model = get_product_name ();
+        string manufacturer = get_sys_vendor ();
+
+        if (manufacturer.length == 0) {
+            return null;
+        }
+        manufacturer = manufacturer.down ();
+
+        if (manufacturer.contains ("to be filled")) {
+            // Don't bother with products in development.
+            return null;
+        }
+
+        if (manufacturer.contains ("bochs") || manufacturer.contains ("vmware")) {
+            model = "virtual machine";
+            // VirtualBox sets an appropriate system-product-name.
+        } else {
+            if (manufacturer.contains ("lenovo") || manufacturer.contains ("ibm")) {
+                model = get_product_version ();
+            }
+        }
+
+        try {
+            if (manufacturer.contains ("apple")) {
+                //  MacBook4,1 - strip the 4,1
+                var re = new Regex ("[^a-zA-Z\\s]");
+                model = re.replace (model, model.length, 0, "");
+            }
+
+            // Replace each gap of non-alphanumeric characters with a dash.
+            // Ensure the resulting string does not begin or end with a dash.
+            var re = new Regex ("[^a-zA-Z0-9]+");
+            model = re.replace (model, model.length, 0, "-");
+            while (model[0] == '-') {
+                model = model.substring (1);
+            }
+            while (model[model.length - 1] == '-') {
+                model = model.substring (0, model.length - 1);
+            }
+
+            if (model.down () == "not-available") {
```

Comment:
The general convention for hostnames is to keep them always lower case anyway. So I think it makes more sense to make sure all of the strings that can form `model` are already lower case, rather than always doing `model.down ()` for comparisons.

Reply:
@danrabbit @cassidyjames do we want to lowercase hostnames? This would mean that, for example, `Star-LabTop-Mk-IV` would become `star-labtop-mk-iv`.

Reply:
I think if uppercase is valid, we should keep them; I do frequently use uppercase in my hostnames just fine 😉